packages/ozone/src/set/service.ts

```diff
@@ -0,0 +1,203 @@
+import Database from '../db'
+import { Selectable, sql } from 'kysely'
+import { OzoneSet } from '../db/schema/ozone_set'
+import { SetView } from '../lexicon/types/tools/ozone/sets/defs'
+
+export type SetServiceCreator = (db: Database) => SetService
+
+export class SetService {
+  constructor(public db: Database) {}
+
+  static creator() {
+    return (db: Database) => new SetService(db)
+  }
+
+  async query({
+    limit,
+    cursor,
+    namePrefix,
+    sortBy,
+  }: {
+    limit: number
+    cursor?: string
+    namePrefix?: string
+    sortBy: 'name' | 'createdAt' | 'updatedAt'
+  }): Promise<{
+    sets: Selectable<OzoneSet & { setSize: number }>[]
+    cursor?: string
+  }> {
+    let qb = this.db.db
+      .selectFrom('ozone_set as s')
+      .leftJoin(
+        (eb) =>
+          eb
+            .selectFrom('ozone_set_value')
+            .select(['setId'])
+            .select((eb) => eb.fn.count('id').as('count'))
+            .groupBy('setId')
+            .as('sv'),
+        (join) => join.onRef('sv.setId', '=', 's.id'),
+      )
+      .select([
+        's.id',
+        's.name',
+        's.description',
+        's.createdAt',
+        's.updatedAt',
+        sql<number>`coalesce(sv.count, 0)`.as('setSize'),
+      ])
+      .limit(limit)
+
+    if (namePrefix) {
+      qb = qb.where('s.name', 'like', `${namePrefix}%`)
+    }
+
+    if (cursor) {
+      if (sortBy === 'name') {
+        qb = qb.where('s.name', '>', cursor)
+      } else {
+        qb = qb.where(`s.${sortBy}`, '>', new Date(cursor))
+      }
+    }
+
+    qb = qb.orderBy(`s.${sortBy}`, 'asc')
+
+    const sets = await qb.execute()
+    const lastItem = sets.at(-1)
+
+    return {
+      sets,
+      cursor: lastItem
+        ? sortBy === 'name'
+          ? lastItem?.name
+          : lastItem?.[sortBy].toISOString()
+        : undefined,
+    }
+  }
+
+  async getByName(
+    name: string,
+    txn: Database,
+  ): Promise<Selectable<OzoneSet> | undefined> {
+    const query = txn.db
+      .selectFrom('ozone_set')
+      .selectAll()
+      .where('name', '=', name)
+
+    return await query.executeTakeFirst()
+  }
+
+  async getSetWithValues({
+    name,
+    limit,
+    cursor,
+  }: {
+    name: string
+    limit: number
+    cursor?: string
+  }): Promise<
+    { set: Selectable<OzoneSet>; values: string[]; cursor?: string } | undefined
+  > {
+    const set = await this.getByName(name, this.db)
+    if (!set) return undefined
+
+    let qb = this.db.db
+      .selectFrom('ozone_set_value')
+      .select(['value'])
+      .where('setId', '=', set.id)
+      .limit(limit)
+
+    if (cursor) {
+      qb = qb.where('value', '>', cursor)
+    }
+
+    qb = qb.orderBy('value', 'asc')
```

Comment:
We can paginate over `value`, though it does have the property that if you're trying to sync the whole set you can miss values that are added while you're paginating.  It might be natural to page over creation time and break ties with the item `id`, so if items are added while you're paginating you'll only encounter them on the final pages.